### PR TITLE
Allow overriding of volume mount path and added a flag to preserve case sensitivity of secrets

### PR DESCRIPTION
**Pull request by smitthakkar96**
This PR adds two new annotations:
- `vault.hashicorp.com/secret-volume-path`
- `vault.hashicorp.com/preserve-secret-case`
- Fixes broken test (it was always passing): [TestSecretAnnotations](https://github.com/hashicorp/vault-k8s/blob/master/agent-inject/agent/annotations_test.go#L107)

Why?
- Currently vault-injector doesn't allow overriding the path where it mounts the secrets. In our case our base image expects the secrets to be present in a particular directory so it gets picked up as env vars when the container starts. 

- Currently vault-injector assumes that the case of the secret name doesn't matter well in our case it is very important, so to not break backwards compatibility and also fullfil the need I added `vault.hashicorp.com/preserve-secret-case`

### CODE REVIEW

File: agent-inject/agent/container_volume.go [hunk 1]
```diff
@@ -8,7 +8,6 @@ const (
 	configVolumeName    = "vault-config"
 	configVolumePath    = "/vault/configs"
 	secretVolumeName    = "vault-secrets"
-	secretVolumePath    = "/vault/secrets"
```
Comment (jasonodonnell) [inline agent-inject/agent/container_volume.go]: This constant should remain and used if `AnnotationVaultSecretVolumePath` is not set.

File: agent-inject/agent/annotations.go [hunk 2]
```diff
@@ -112,6 +112,12 @@ const (
 	// AnnotationVaultAuthPath specifies the mount path to be used for the Kubernetes auto-auth
 	// method.
 	AnnotationVaultAuthPath = "vault.hashicorp.com/auth-path"
+
+	// AnnotationVaultSecretVolumePath specifies where the secrets are to be
+	// Mounted after fetching.
+	AnnotationVaultSecretVolumePath = "vault.hashicorp.com/secret-volume-path"
+
+	AnnotationPreserveSecretCase = "vault.hashicorp.com/preserve-secret-case"
```
Comment (jasonodonnell) [inline agent-inject/agent/annotations.go]: Needs comment for go doc.

File: Makefile [hunk 3]
```diff
@@ -1,6 +1,6 @@
 REGISTRY_NAME?=docker.io/hashicorp
 IMAGE_NAME=vault-k8s
-VERSION?=0.2.0
+VERSION?=0.2.1
```
Comment (jasonodonnell) [inline Makefile]: Please change this back to `0.2.0`, we'll handle release increments.

File: agent-inject/agent/annotations.go [hunk 4]
```diff
@@ -173,6 +179,13 @@ func Init(pod *corev1.Pod, image, address, authPath, namespace string) error {
 		pod.ObjectMeta.Annotations[AnnotationAgentRequestsMem] = DefaultResourceRequestMem
 	}
 
+	if _, ok := pod.ObjectMeta.Annotations[AnnotationPreserveSecretCase]; !ok {
+		pod.ObjectMeta.Annotations[AnnotationPreserveSecretCase] = "false"
+	}
+	if _, ok := pod.ObjectMeta.Annotations[AnnotationVaultSecretVolumePath]; !ok {
+		pod.ObjectMeta.Annotations[AnnotationVaultSecretVolumePath] = "/vault/secrets"
```
Comment (jasonodonnell) [inline agent-inject/agent/annotations.go]: This should use the secret path constant instead.

File: agent-inject/agent/annotations.go [hunk 5]
```diff
@@ -173,6 +179,13 @@ func Init(pod *corev1.Pod, image, address, authPath, namespace string) error {
 		pod.ObjectMeta.Annotations[AnnotationAgentRequestsMem] = DefaultResourceRequestMem
 	}
 
+	if _, ok := pod.ObjectMeta.Annotations[AnnotationPreserveSecretCase]; !ok {
+		pod.ObjectMeta.Annotations[AnnotationPreserveSecretCase] = "false"
```
Comment (jasonodonnell) [inline agent-inject/agent/annotations.go]: This should be a bool to be consistent with other annotations: https://github.com/hashicorp/vault-k8s/blob/master/agent-inject/agent/annotations.go#L210-L244

File: agent-inject/agent/annotations.go [hunk 6]
```diff
@@ -188,7 +201,12 @@ func secrets(annotations map[string]string) []*Secret {
 		secretName := fmt.Sprintf("%s-", AnnotationAgentInjectSecret)
 		if strings.Contains(name, secretName) {
 			raw := strings.ReplaceAll(name, secretName, "")
-			name := strings.ToLower(raw)
+			var name string
+			if annotations[AnnotationPreserveSecretCase] == "false" {
```
Comment (jasonodonnell) [inline agent-inject/agent/annotations.go]: Should use boolean value here.

**Comment by smitthakkar96** [comment]
Thanks @jasonodonnell will address the comments

**Comment by smitthakkar96** [comment]
@jasonodonnell I addressed the comments

**Comment by smitthakkar96** [comment]
@jasonodonnell any updates?

**Comment by ShabihZaidi** [comment]
any update, I want this feature too

**Comment by jasonodonnell** [comment]
Hi @smitthakkar96, sorry for the delay, but I'm ready to review this now!  When you get a chance can you resolve the conflicts?  Thanks!

**Comment by smitthakkar96** [comment]
@jasonodonnell resolved

File: agent-inject/agent/annotations_test.go [hunk 7]
```diff
@@ -135,20 +135,60 @@ func TestSecretAnnotations(t *testing.T) {
...
-				}
 			}
 		} else if len(agent.Secrets) > 0 {
 			t.Error("Secrets length was greater than zero, it shouldn't have been")
 		}
+	}
+}
+
+func TestSecretAnnotationsWithPreserveCaseSensitivityFlagOn(t *testing.T) {
+	tests := []struct {
```
Comment (smitthakkar96) [inline agent-inject/agent/annotations_test.go]: @jasonodonnell this is not dry but making it dry would make tests a bit complicated so I duplicated the code

**Comment by smitthakkar96** [comment]
ping!

**Comment by smitthakkar96** [comment]
@jasonodonnell isn't it better to still keep it global and allow overriding at secret level if needed? I am happy to update the code to allow overriding the path and case sensitivity flag at secret level if we agree

**Comment by saurav-k** [comment]
@smitthakkar96 @jasonodonnell   Your comments make more sense to have path  configurable per secret. Specially with pattern you suggested .

**Comment by smitthakkar96** [comment]
@saurav-k I still think having a global flag makes it easy for people to configure incase if you want to override the path for all secrets you are fetching. Incase of if you want to override per secret you can still do if we implement what @jasonodonnell is suggesting

**Comment by saurav-k** [comment]
@smitthakkar96  Makes sense to have a global flag in case you want to mount all secret at same path and  this PR can be scoped to global flag and another PR can be created for the feature suggested by @jasonodonnell   , My thoughts on suggested feature. I have seen apps need files from different mount location as secret  . and in that case suggested feature will be very useful. Let's wait for @jasonodonnell  to give his view on scope of this PR.

**Comment by jasonodonnell** [comment]
@smitthakkar96 I agree the global flag is also valuable because in some situations it would eliminate a lot of annotations. It would likely be that secret specific configurations override the globals.

**Comment by smitthakkar96** [comment]
Ok I will work on it tonight and hopefully update the PR.

Best Regards,
Smit

**Comment by smitthakkar96** [comment]
@jasonodonnell I made the changes, please have a look when you get chance

**Comment by smitthakkar96** [comment]
@jasonodonnell thanks for merging it :)

**Pull request merged**